merkle.go

```diff
@@ -0,0 +1,189 @@
+				sourceData: [][]byte{
+					nodes[i].Hash,
+					nodes[i+1].Hash,
+				},
+			})
+		}
+
+		// The Merkle tree is being built from bottom to top,
+		// so each level has exactly 1/2 fewer nodes
+		// than the previous level (property of perfect binary trees).
+		// Therefore, for N nodes on a single tree level, only N/2 results can be expected
+		for i := 0; i < len(nodes)/2; i++ {
+			result := workerPool.getResult()
+			if result.error != nil {
+				return nil, fmt.Errorf(
+					"unable to perform hashing, %w",
+					result.error,
+				)
+			}
+
+			// Create a placeholder for the parent node
+			parent := &Node{
+				// Save the hashing data of the 2 children
+				Hash: result.hashData,
+				// Save a reference to the left child
+				Left: nodes[result.storeIndex],
+				// Save a reference to the right child
+				Right: nodes[result.storeIndex+1],
+			}
+
+			// Save the parent reference with the children
+			nodes[result.storeIndex].Parent = parent
+			nodes[result.storeIndex+1].Parent = parent
+
+			// Overwrite the left child's slot in the array,
+			// since it's no longer needed. The right child
+			// is also not needed anymore in the original array,
+			// and will be overwritten later
+			nodes[result.storeIndex] = parent
+		}
+
+		// Now that results are gathered for the level,
+		// the array can be shifted and shrunk
+		shiftAndShrinkArray(&nodes)
+	}
+
+	return &MerkleTree{
+		Root: nodes[0],
+	}, nil
+}
+
+// isValidDataSet makes sure the input set has elements
+func isValidDataSet(inputData [][]byte) bool {
+	return inputData != nil && len(inputData) > 0
+}
+
+// shiftAndShrinkArray shifts every other node to the
+// beginning of the array, and discards half of it (shrinks it).
+// Due to the way results are being stored (index of left child),
+// and the fact that the Merkle tree is a perfect binary tree,
+// it can be guaranteed that the results are on every other index in the node level array
+func shiftAndShrinkArray(nodes *[]*Node) {
+	// Put the results in the first half of the array.
+	// One counter keeps track of the next slot to place the value (moves by 1) (saveIndx)
+	// and the other keeps track of which element should be stored (moves by 2) (resultIndx)
+	initialLevelSize := len(*nodes)
+	saveIndx := 0
+	for resultIndx := 0; resultIndx < initialLevelSize; resultIndx += 2 {
+		(*nodes)[saveIndx] = (*nodes)[resultIndx]
+		saveIndx++
+	}
+
+	// Wipe the other half of the array, since
+	// all useful and needed results are in the first half
+	*nodes = (*nodes)[:initialLevelSize/2]
+}
+
+// generateLeaves generates the initial (leaf) level of the Merkle tree.
+// The leaf level needs to be a power of 2, since the Merkle tree is considered
+// to be a perfect binary tree
+func generateLeaves(inputData [][]byte, wp *workerPool) ([]*Node, error) {
+	inputDataSize := len(inputData)
+	leafLevelSize := inputDataSize
+
+	if inputDataSize > 1 {
+		// Find the nearest power of 2 for the base leaf level size
```

Comment:
Why `leafLevelSize` must be 2^n?
According to https://bitcoin.stackexchange.com/questions/79364/are-number-of-transactions-in-merkle-tree-always-even, it's fine the number of elements is even

Reply:
They are different. I can show you simple example.

We're creating Merkle tree from A, B, C, D, E, F. But they are only 6 elements. There are 2 ways we can choose.
(1) When you find lack of node in the opposite side, copy one side to another side, and calculate hash
(2) Filled out with leaves so that number of leaves is 2^n (copy last element) firstly

As below image, they return different values

![Screenshot 2022-03-28 at 19 55 09](https://user-images.githubusercontent.com/19720977/160448612-54a87c2d-875b-45e2-9c62-365667aec109.png)

btw, according to https://github.com/cbergoon/merkletree, it's choosing (1)

Reply:
How do you suggest we proceed? (1)?

Reply:
I'm guessing the solution here is to just duplicate the last element if we see the input array for the tasks is odd.
So the flow would look something like this:

* Prepare an **even** array for the input tasks
* Make jobs with the even array
* Gather results the same way we do now, where we safe all relevant data in the first half of the array
* Prune the other half of the array
* If the leftover array is odd, duplicate the last element
* Repeat 🙂